File: node/packages/core/src/text-revision.ts
// FILE: node/packages/core/src/text-revision.ts
//
// Whole-text revision: diff the document's CLEAN view against the text the
// caller supplies, turn the difference into tracked changes, and refuse to
// hand back a document whose clean view does not then read exactly like that
// text. Port of python/src/adeu/text_revision.py minus its CLI-only parts.
//
// DELIBERATE structural difference from the Python twin: this module does no
// file IO. It takes a loaded document and returns bytes; the caller (the MCP
// server) already owns reading, writing, and the agent-facing shape of every
// filesystem error. The verification failure therefore travels as an exception
// CARRYING the diagnostic copy's bytes and its intended path — the handler
// writes it — instead of writing the sibling itself.
import { basename, dirname, extname, join } from "node:path";

import { DocumentObject } from "./docx/bridge.js";
import { generate_edits_via_paragraph_alignment } from "./diff.js";
import { _extractTextFromDoc } from "./ingest.js";
import { RedlineEngine } from "./engine.js";
import type { ModifyText } from "./models.js";

// Only the OPEN tokens: a bare closing token is ordinary prose far more often
// than it is markup ("A ~> B", "rate++}"), and markup view never emits one
// without its opener (verifier finding, Task 15 attempt 3).
const _CRITICMARKUP_TOKENS = ["{++", "{--", "{~~", "{==", "{>>"];

const _RE_FILE_PATH_BANNER = /^> \*\*File Path:\*\*[^\n]*\n+/;
const _RE_PAGE_BANNER = /^> \*\*Page (\d+) of (\d+)\*\*[^\n]*\n+(?:---\n+)?/;
const _RE_PAGE_FOOTER =
  /\n+---\n+> \*\*Continues on page (\d+) of (\d+)\.\*\*[^\n]*\s*$/;
const _RE_APPENDIX_POINTER = /\n+---\n+> \*\*Appendix available\.\*\*[^\n]*\s*$/;

// Documents at or above this many characters use the 50% deletion budget;
// shorter ones use the higher 75% floor (see check_major_deletions).
const _MAJOR_DELETION_MIN_ORIGINAL_CHARS = 2000;

/** Base error for every text-revision refusal (Python raises ValueError). */
export class TextRevisionError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "TextRevisionError";
  }
}

/** Raised when clean-text post-apply verification fails. */
export class TextRevisionVerificationError extends TextRevisionError {
  constructor(
    message: string,
    public unverified_path: string,
    public output_path: string,
    public stats: Record<string, any>,
    public unverified_bytes: Uint8Array,
  ) {
    super(message);
    this.name = "TextRevisionVerificationError";
  }
}

/**
 * The effective Track Changes author: explicit name, else `ADEU_AUTHOR`, else
 * the Node engine's own default.
 *
 * Python additionally falls back to the OS account name (`getpass.getuser()`,
 * filtered against machine accounts) because its CLI has a human user. This
 * server does not, so the last resort is the engine default the Node MCP
 * schema already advertises — a declared cosmetic difference, recorded in
 * shared/conformance/README.md.
 */
function get_default_author(author?: string | null): string {
  if (author && author.trim()) return author.trim();
  const env_author = (process.env.ADEU_AUTHOR || "").trim();
  if (env_author) return env_author;
  return "Adeu AI (TS)";
}

/** Refuses revised text input if it contains CriticMarkup syntax. */
export function check_criticmarkup(text: string): void {
  if (_CRITICMARKUP_TOKENS.some((tok) => text.includes(tok))) {
    throw new TextRevisionError(
      "Revised text contains CriticMarkup tokens ({++..++}, {--..--}, {~~..~>..~~}, {==..==}, " +
        "{>>..<<}). `apply_text_revision` compares text against the document's CLEAN view, " +
        "so CriticMarkup tokens would be diffed into the document as literal prose.",
    );
  }
}

/** Strips extract header, banners, and footers from text input. */
export function strip_page_chrome(text: string): {
  text: string;
  page: number | null;
  total: number | null;
} {
  text = text.replace(_RE_FILE_PATH_BANNER, "");
  let page: number | null = null;
  let total: number | null = null;
  const banner = _RE_PAGE_BANNER.exec(text);
  if (banner) {
    page = parseInt(banner[1], 10);
    total = parseInt(banner[2], 10);
    text = text.slice(banner[0].length);
  }
  text = text.replace(_RE_APPENDIX_POINTER, "");
  const footer = _RE_PAGE_FOOTER.exec(text);
  if (footer) {
    if (page === null) page = parseInt(footer[1], 10) - 1;
    if (total === null) total = parseInt(footer[2], 10);
    text = text.slice(0, footer.index);
  }
  return { text, page, total };
}

/**
 * Refuses to silently delete the majority of a document. 2000 chars ≈ one
 * page of prose; above that, losing half the document is almost never
 * intentional. Short documents matter too (QA 2026-07-19 v8 F-12): below the
 * threshold the guard still arms, at a higher 75% floor so that deliberately
 * halving a small draft stays a one-command workflow while near-total
 * truncation requires the explicit flag.
 *
 * The budget is measured in CHARACTERS only: a document made of many short
 * paragraphs legitimately loses dozens of them in an ordinary edit.
 * `source_name` names the revised text's origin (the CLI's text file) in the
 * refusal message when there is one.
 */
export function check_major_deletions(
  original_text: string,
  revised_text: string,
  allow_major_deletions: boolean = false,
  source_name?: string | null,
): void {
  if (allow_major_deletions) return;

  const orig_len = original_text.length;
  const rev_len = revised_text.length;
  if (orig_len === 0) return;

  const char_deletion_ratio = (orig_len - rev_len) / orig_len;
  const threshold =
    orig_len >= _MAJOR_DELETION_MIN_ORIGINAL_CHARS ? 0.5 : 0.75;
  if (char_deletion_ratio <= threshold) return;

  const subject = source_name ? `'${source_name}'` : "The revised text";
  throw new TextRevisionError(
    `${subject} is ~${Math.floor(char_deletion_ratio * 100)}% shorter than the document's clean text ` +
      `(${rev_len.toLocaleString("en-US")} vs ${orig_len.toLocaleString("en-US")} characters, ` +
      `threshold is >${Math.floor(threshold * 100)}% deletion). ` +
      "Applying it would delete the majority of the document as tracked deletions.\n" +
      "   If the text is a partial extract, re-extract the ENTIRE document with " +
      "`--page all --clean-view` and edit that.\n" +
      "   If the mass deletion is intentional, re-run with --allow-major-deletions " +
      "(over MCP: allow_major_deletions=True).",
  );
}

/** Strips synthetic {#cell:<paraId>} anchor tokens from clean-text payloads. */
export function strip_cell_anchors(text: string): string {
  text = text.replace(/([^|])\s+\{#cell:[^}]+\}/g, "$1");
  text = text.replace(/\{#cell:[^}]+\}/g, "");
  return text;
}

/** Extracts clean accepted text from a document (no generated appendix). */
function _extract_clean_text_from_doc(doc: DocumentObject): string {
  return _extractTextFromDoc(doc, true, false) as string;
}

/** Normalizes Markdown heading chrome and synthetic anchors for clean-text verification. */
function _normalize_virtual_projection_text(text: string): string {
  return strip_cell_anchors(text.replace(/^#+\s*/gm, ""));
}

// Python's `str.isprintable()` is False for every code point in category Other
// (C) or Separator (Z), the ASCII space alone excepted — so this is the exact
// set `repr()` escapes.
const _RE_NONPRINTABLE = /\p{C}|\p{Z}/u;

/**
 * Python's `repr()` of a short string, for the divergence message. NOT
 * `JSON.stringify`: Python prefers single quotes and only switches to double
 * quotes when the text holds a single quote but no double quote, and it sizes
 * each escape to the code point (`\xNN`, `\uNNNN`, `\UNNNNNNNN`) rather than
 * always emitting `\uNNNN`. The two engines' failure text has to read
 * identically (shared/conformance/README.md), and this string is quoted from a
 * real document, so tabs, breaks (`\x0b`) and apostrophes all turn up.
 *
 * The escape set is the FULL non-printable set, not just C0/C1 (verifier
 * finding, Task 19 attempt 2): NBSP, soft hyphen and zero-width space are
 * everyday Word characters, and left raw they render the message useless —
 * both excerpts read "Fee 1000" while differing by an invisible code point,
 * which is the one thing quoting the excerpts is for.
 */
function _repr(text: string): string {
  const quote = text.includes("'") && !text.includes('"') ? '"' : "'";
  let body = "";
  // By code point, so astral characters are escaped whole, not per surrogate.
  for (const ch of text) {
    if (ch === "\\") body += "\\\\";
    else if (ch === "\n") body += "\\n";
    else if (ch === "\r") body += "\\r";
    else if (ch === "\t") body += "\\t";
    else if (ch === quote) body += `\\${quote}`;
    else if (ch !== " " && _RE_NONPRINTABLE.test(ch)) {
      const cp = ch.codePointAt(0)!;
      const [prefix, width] =
        cp < 0x100 ? ["x", 2] : cp < 0x10000 ? ["u", 4] : ["U", 8];
      body += `\\${prefix}${cp.toString(16).padStart(width, "0")}`;
    } else body += ch;
  }
  return `${quote}${body}${quote}`;
}

/** Verifies that the clean view of `doc` matches `expected_text`. */
export function verify_clean_text(
  doc: DocumentObject,
  expected_text: string,
): [boolean, string | null] {
  const actual_clean = _extract_clean_text_from_doc(doc);
  const actual_norm = _normalize_virtual_projection_text(actual_clean.trim());
  const expected_norm = _normalize_virtual_projection_text(expected_text.trim());

  if (actual_norm !== expected_norm) {
    // BY CODE POINT, like Python's `zip(actual_norm, expected_norm)` and
    // `norm[div : div + 40]` (verifier finding, Task 19 attempt 3): indexing
    // UTF-16 units instead halves a surrogate pair straddling the 40-character
    // window, and `_repr` then quotes a lone `\ud83d` into the one excerpt this
    // message exists to show. `div` is therefore a code-point index too —
    // Python's number, and nothing else consumes it.
    const actual_cps = Array.from(actual_norm);
    const expected_cps = Array.from(expected_norm);
    let div = 0;
    const minLen = Math.min(actual_cps.length, expected_cps.length);
    while (div < minLen && actual_cps[div] === expected_cps[div]) div++;

    const actual_slice = actual_cps.slice(div, div + 40).join("");
    const expected_slice = expected_cps.slice(div, div + 40).join("");

    const msg =
      "Post-apply verification failed: the applied document's clean text does not match " +
      `the supplied text (first divergence at character ${div}: ` +
      `applied reads ${_repr(actual_slice)}, supplied text reads ` +
      `${_repr(expected_slice)}). The document structure could not fully realize ` +
      "the requested text (e.g. headings or table cells cannot be deleted via text replacement).";
    return [false, msg];
  }
  return [true, null];
}

/** `x.docx` → `x_redlined.docx`; an already-suffixed artifact stays in place. */
export function _default_output_path(input_path: string): string {
  const dir = dirname(input_path);
  const base = basename(input_path);
  const dot = base.lastIndexOf(".");
  const stem = dot === -1 ? base : base.slice(0, dot);
  if (stem.endsWith("_redlined") || stem.endsWith("_processed")) {
    return input_path;
  }
  return join(dir, `${stem}_redlined.docx`);
}

/** `<target stem>.unverified.docx`, the diagnostic sibling's path. */
function _unverified_path(output_path: string): string {
  const stem = basename(output_path, extname(output_path));
  return join(dirname(output_path), `${stem}.unverified.docx`);
}

/**
 * Core whole-text diff→tracked-changes primitive with a clean-text
 * verification gate. Returns the bytes to write; throws
 * TextRevisionVerificationError (carrying the diagnostic copy) when the
 * applied document cannot be proven to match `revised_text`.
 */
export async function apply_text_revision_core(opts: {
  doc: DocumentObject;
  input_path: string;
  revised_text: string;
  output_path?: string | null;
  author?: string | null;
  allow_major_deletions?: boolean;
}): Promise<{
  stats: Record<string, any>;
  output_path: string;
  out_bytes: Uint8Array;
  unverified?: { path: string; bytes: Uint8Array };
}> {
  const { doc, input_path, revised_text } = opts;
  const { text: raw_text_clean, page, total } = strip_page_chrome(revised_text);
  const text_clean_input = strip_cell_anchors(raw_text_clean);
  if (total !== null && total > 1) {
    throw new TextRevisionError(
      `Text revision looks like page ${page || "?"} of ${total} of a paginated extract — ` +
        "it contains only part of the document, and applying it would delete every page " +
        "not present. Re-extract the ENTIRE document first with --page all --clean-view.",
    );
  }

  check_criticmarkup(text_clean_input);

  const text_orig = _extract_clean_text_from_doc(doc);

  check_major_deletions(
    text_orig,
    text_clean_input,
    opts.allow_major_deletions ?? false,
  );

  const changes: ModifyText[] = generate_edits_via_paragraph_alignment(
    text_orig,
    text_clean_input,
  );
  const engine = new RedlineEngine(doc, get_default_author(opts.author));
  const stats = engine.process_batch(changes as any);

  const target_output = opts.output_path
    ? opts.output_path
    : _default_output_path(input_path);

  const [verified, err_msg] = verify_clean_text(engine.doc, text_clean_input);

  if (!verified) {
    const unverified_path = _unverified_path(target_output);
    const unverified_bytes = await engine.doc.save();

    const full_err =
      `${err_msg} Nothing was written to '${target_output}'; a diagnostic copy was kept ` +
      `at '${unverified_path}' — it is NOT the requested document.`;
    stats.verified = false;
    stats.verification_error = full_err;
    stats.error = "verification_failed";
    stats.edits_skipped = (stats.edits_applied || 0) + (stats.edits_skipped || 0);
    stats.edits_applied = 0;
    stats.actions_skipped =
      (stats.actions_applied || 0) + (stats.actions_skipped || 0);
    stats.actions_applied = 0;
    stats.output_path = null;
    stats.unverified_output_path = unverified_path;
    if (stats.edits) {
      for (const report of stats.edits) {
        report.status = "failed";
        report.error = "Not applied: post-apply verification failed.";
        report.critic_markup = null;
        report.clean_text = null;
      }
    }

    throw new TextRevisionVerificationError(
      full_err,
      unverified_path,
      target_output,
      stats,
      unverified_bytes,
    );
  }

  const out_bytes = await engine.doc.save();
  stats.output_path = target_output;
  stats.verified = true;

  return { stats, output_path: target_output, out_bytes };
}

File: python/src/adeu/text_revision.py
import os
import re
from io import BytesIO
from pathlib import Path
from typing import Any, Dict, Optional, Tuple, Union

from adeu.diff import generate_edits_via_paragraph_alignment
from adeu.ingest import _extract_text_from_doc
from adeu.models import DocumentChange
from adeu.redline.engine import RedlineEngine
from adeu.utils.docx import strip_bom_from_docx_bytes
from adeu.utils.opc import load_document as Document

# Only the OPEN tokens: a bare closing token is ordinary prose far more often
# than it is markup ("A ~> B", "rate++}"), and markup view never emits one
# without its opener (verifier finding, Task 15 attempt 3).
_CRITICMARKUP_TOKENS = ("{++", "{--", "{~~", "{==", "{>>")

_EXTRACT_HEADER_RE = re.compile(r"^> \*\*File Path:\*\*[^\n]*\n+")
_PAGE_BANNER_RE = re.compile(r"^> \*\*Page (\d+) of (\d+)\*\*[^\n]*\n+(?:---\n+)?")
_PAGE_FOOTER_RE = re.compile(r"\n+---\n+> \*\*Continues on page (\d+) of (\d+)\.\*\*[^\n]*\s*$")
_APPENDIX_POINTER_RE = re.compile(r"\n+---\n+> \*\*Appendix available\.\*\*[^\n]*\s*$")

_MACHINE_ACCOUNT_NAMES = {"root", "admin", "administrator", "system", "daemon", "nobody"}

# Documents at or above this many characters use the 50% deletion budget;
# shorter ones use the higher 75% floor (see check_major_deletions).
_MAJOR_DELETION_MIN_ORIGINAL_CHARS = 2000


class TextRevisionError(Exception):
    """Base exception for text revision errors."""

    pass


class TextRevisionVerificationError(TextRevisionError):
    """Raised when clean-text post-apply verification fails."""

    def __init__(self, message: str, unverified_path: Path, output_path: Path, stats: Optional[Dict[str, Any]] = None):
        super().__init__(message)
        self.unverified_path = unverified_path
        self.output_path = output_path
        self.stats = stats or {}


def get_default_author(author: Optional[str] = None) -> str:
    """Returns the effective author name for track changes."""
    if author and author.strip():
        return author.strip()
    env_author = (os.environ.get("ADEU_AUTHOR") or "").strip()
    if env_author:
        return env_author
    try:
        import getpass

        user = getpass.getuser()
    except Exception:
        return "Adeu AI"
    if not user or user.strip().lower() in _MACHINE_ACCOUNT_NAMES:
        return "Adeu AI"
    return user


def check_criticmarkup(text: str) -> None:
    """Refuses revised text input if it contains CriticMarkup syntax."""
    if any(tok in text for tok in _CRITICMARKUP_TOKENS):
        raise ValueError(
            "Revised text contains CriticMarkup tokens ({++..++}, {--..--}, {~~..~>..~~}, {==..==}, "
            "{>>..<<}). `apply_text_revision` compares text against the document's CLEAN view, "
            "so CriticMarkup tokens would be diffed into the document as literal prose."
        )


def _strip_page_chrome(text: str) -> Tuple[str, Optional[int], Optional[int]]:
    """Strips extract header, banners, and footers from text input."""
    text = _EXTRACT_HEADER_RE.sub("", text, count=1)
    page = total = None
    banner = _PAGE_BANNER_RE.match(text)
    if banner:
        page, total = int(banner.group(1)), int(banner.group(2))
        text = text[banner.end() :]
    text = _APPENDIX_POINTER_RE.sub("", text)
    footer = _PAGE_FOOTER_RE.search(text)
    if footer:
        if page is None:
            page = int(footer.group(1)) - 1
        if total is None:
            total = int(footer.group(2))
        text = text[: footer.start()]
    return text, page, total


def check_major_deletions(
    original_text: str,
    revised_text: str,
    allow_major_deletions: bool = False,
    source_name: Optional[str] = None,
) -> None:
    """
    Refuses to silently delete the majority of a document. 2000 chars ≈ one
    page of prose; above that, losing half the document is almost never
    intentional. Short documents matter too (QA 2026-07-19 v8 F-12): below the
    threshold the guard still arms, at a higher 75% floor so that deliberately
    halving a small draft stays a one-command workflow while near-total
    truncation requires the explicit flag.

    The budget is measured in CHARACTERS only: a document made of many short
    paragraphs legitimately loses dozens of them in an ordinary edit.
    `source_name` names the revised text's origin (the CLI's text file) in the
    refusal message when there is one.
    """
    if allow_major_deletions:
        return

    orig_len = len(original_text)
    rev_len = len(revised_text)
    if orig_len == 0:
        return

    char_deletion_ratio = (orig_len - rev_len) / orig_len
    threshold = 0.50 if orig_len >= _MAJOR_DELETION_MIN_ORIGINAL_CHARS else 0.75
    if char_deletion_ratio <= threshold:
        return

    subject = f"'{source_name}'" if source_name else "The revised text"
    raise ValueError(
        f"{subject} is ~{int(char_deletion_ratio * 100)}% shorter than the document's clean text "
        f"({rev_len:,} vs {orig_len:,} characters, threshold is >{int(threshold * 100)}% deletion). "
        "Applying it would delete the majority of the document as tracked deletions.\n"
        "   If the text is a partial extract, re-extract the ENTIRE document with "
        "`--page all --clean-view` and edit that.\n"
        "   If the mass deletion is intentional, re-run with --allow-major-deletions "
        "(over MCP: allow_major_deletions=True)."
    )


def _extract_clean_text_from_doc(doc: Any) -> str:
    """Extracts clean accepted text from a DOCX Document object."""
    res = _extract_text_from_doc(doc, clean_view=True, include_appendix=False)
    if isinstance(res, tuple):
        return str(res[0])
    return str(res)


def strip_cell_anchors(text: str) -> str:
    """Strips synthetic {#cell:<paraId>} anchor tokens from clean-text payloads."""
    text = re.sub(r"([^|])\s+\{#cell:[^}]+\}", r"\1", text)
    text = re.sub(r"\{#cell:[^}]+\}", "", text)
    return text


def _normalize_virtual_projection_text(text: str) -> str:
    """Normalizes Markdown heading chrome and synthetic anchors for clean-text verification."""
    text = re.sub(r"^#+\s*", "", text, flags=re.MULTILINE)
    return strip_cell_anchors(text)


def verify_clean_text(
    engine_doc: Any,
    expected_text: str,
) -> Tuple[bool, Optional[str]]:
    """Verifies that the clean view of engine_doc matches expected_text."""
    actual_clean = _extract_clean_text_from_doc(engine_doc)
    expected = expected_text.strip()
    actual = actual_clean.strip()

    actual_norm = _normalize_virtual_projection_text(actual)
    expected_norm = _normalize_virtual_projection_text(expected)

    if actual_norm != expected_norm:
        div = next(
            (k for k, (a, b) in enumerate(zip(actual_norm, expected_norm, strict=False)) if a != b),
            min(len(actual_norm), len(expected_norm)),
        )
        msg = (
            "Post-apply verification failed: the applied document's clean text does not match "
            f"the supplied text (first divergence at character {div}: "
            f"applied reads {actual_norm[div : div + 40]!r}, supplied text reads "
            f"{expected_norm[div : div + 40]!r}). The document structure could not fully realize "
            "the requested text (e.g. headings or table cells cannot be deleted via text replacement)."
        )
        return False, msg
    return True, None


def _load_docx_from_path(path: Union[str, Path]) -> Tuple[Any, bytes]:
    p = Path(path)
    if not p.exists() or not p.is_file():
        raise FileNotFoundError(f"DOCX file not found: {p}")
    raw_bytes = p.read_bytes()
    sanitized_bytes = strip_bom_from_docx_bytes(raw_bytes)
    doc = Document(BytesIO(sanitized_bytes))
    return doc, sanitized_bytes


def _write_output_bytes(path: Path, data: bytes) -> None:
    import tempfile

    if path.parent and str(path.parent) not in ("", "."):
        path.parent.mkdir(parents=True, exist_ok=True)
    fd, tmp_name = tempfile.mkstemp(prefix=f".{path.name}.", suffix=".tmp", dir=str(path.parent or "."))
    tmp_path = Path(tmp_name)
    try:
        with os.fdopen(fd, "wb") as fb:
            fb.write(data)
        os.replace(tmp_path, path)
    finally:
        if tmp_path.exists():
            try:
                tmp_path.unlink()
            except OSError:
                pass


def apply_text_revision_core(
    file_path: Union[str, Path],
    revised_text: str,
    output_path: Optional[Union[str, Path]] = None,
    author: Optional[str] = None,
    allow_major_deletions: bool = False,
) -> Tuple[Dict[str, Any], Path]:
    """Core whole-text diff->tracked-changes primitive with clean-text verification gate."""
    p_input = Path(file_path)
    raw_text_clean, page, total = _strip_page_chrome(revised_text)
    text_clean_input = strip_cell_anchors(raw_text_clean)
    if total is not None and total > 1:
        raise ValueError(
            f"Text revision looks like page {page or '?'} of {total} of a paginated extract — "
            "it contains only part of the document, and applying it would delete every page "
            "not present. Re-extract the ENTIRE document first with --page all --clean-view."
        )

    check_criticmarkup(text_clean_input)

    doc, doc_bytes = _load_docx_from_path(p_input)
    text_orig = _extract_clean_text_from_doc(doc)

    check_major_deletions(text_orig, text_clean_input, allow_major_deletions=allow_major_deletions)

    changes: list[DocumentChange] = list(generate_edits_via_paragraph_alignment(text_orig, text_clean_input))

    author_name = get_default_author(author)
    engine = RedlineEngine(BytesIO(doc_bytes), author=author_name)
    stats = engine.process_batch(changes)

    target_output: Path
    if output_path:
        target_output = Path(output_path)
    else:
        if p_input.stem.endswith("_redlined") or p_input.stem.endswith("_processed"):
            target_output = p_input
        else:
            target_output = p_input.with_name(f"{p_input.stem}_redlined.docx")

    verified, err_msg = verify_clean_text(engine.doc, text_clean_input)

    if not verified:
        unverified_path = target_output.with_name(f"{target_output.stem}.unverified.docx")
        unverified_bytes = engine.save_to_stream().getvalue()
        _write_output_bytes(unverified_path, unverified_bytes)

        full_err = (
            f"{err_msg} Nothing was written to '{target_output}'; a diagnostic copy was kept "
            f"at '{unverified_path}' — it is NOT the requested document."
        )
        stats["verified"] = False
        stats["verification_error"] = full_err
        stats["error"] = "verification_failed"
        stats["edits_skipped"] = stats.get("edits_applied", 0) + stats.get("edits_skipped", 0)
        stats["edits_applied"] = 0
        stats["actions_skipped"] = stats.get("actions_applied", 0) + stats.get("actions_skipped", 0)
        stats["actions_applied"] = 0
        stats["output_path"] = None
        stats["unverified_output_path"] = str(unverified_path)
        if stats.get("edits"):
            for report in stats["edits"]:
                report["status"] = "failed"
                report["error"] = "Not applied: post-apply verification failed."
                report["critic_markup"] = None
                report["clean_text"] = None

        raise TextRevisionVerificationError(
            full_err, unverified_path=unverified_path, output_path=target_output, stats=stats
        )

    # Save output
    output_bytes = engine.save_to_stream().getvalue()
    _write_output_bytes(target_output, output_bytes)

    stats["output_path"] = str(target_output)
    stats["verified"] = True

    return stats, target_output
